service/controller/resource/terminateunhealthynode/create.go

```diff
@@ -0,0 +1,154 @@
+package terminateunhealthynode
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/autoscaling"
+	"github.com/giantswarm/badnodedetector/pkg/detector"
+	"github.com/giantswarm/kubelock/v2"
+	"github.com/giantswarm/microerror"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	"github.com/giantswarm/aws-operator/pkg/project"
+	"github.com/giantswarm/aws-operator/service/controller/controllercontext"
+	"github.com/giantswarm/aws-operator/service/controller/key"
+)
+
+const (
+	annotationEnableNodeTermination = "node.giantswarm.io/terminate-unhealthy"
+
+	nodeTerminationTickThreshold = 6
+	lockNamespace                = "default"
+	pauseBetweenTermination      = time.Minute * 10
+)
+
+func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+	var err error
+	cr, err := key.ToCluster(ctx, obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	cc, err := controllercontext.FromContext(ctx)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	// check for annotation enabling the node auto repair feature
+	if _, ok := cr.Annotations[annotationEnableNodeTermination]; !ok {
+		r.logger.LogCtx(ctx, "level", "debug", "message", " node auto repair is not enabled for this cluster, cancelling")
+		return nil
+	}
+
+	if cc.Client.TenantCluster.K8s == nil {
+		r.logger.LogCtx(ctx, "level", "debug", "message", "kubernetes clients are not available in controller context yet")
+		r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
+
+		return nil
+	}
+
+	var lock *kubelock.KubeLock
+	{
+		kubelockConfig := kubelock.Config{
+			DynClient: cc.Client.TenantCluster.K8s.DynClient(),
+			GVR: schema.GroupVersionResource{
+				Group:    "",
+				Version:  "v1",
+				Resource: "namespaces",
+			},
+		}
+
+		lock, err = kubelock.New(kubelockConfig)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	acquiredOptions := kubelock.AcquireOptions{
+		TTL: pauseBetweenTermination,
+	}
+
+	var detectorService *detector.Detector
+	{
+		detectorConfig := detector.Config{
+			K8sClient: cc.Client.TenantCluster.K8s.CtrlClient(),
+			Logger:    r.logger,
+
+			NotReadyTickThreshold: nodeTerminationTickThreshold,
+		}
+
+		detectorService, err = detector.NewDetector(detectorConfig)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	nodesToTerminate, err := detectorService.DetectBadNodes(ctx)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	if len(nodesToTerminate) > 0 {
+		err = lock.Lock(project.Name()).Acquire(ctx, lockNamespace, acquiredOptions)
```

Comment:
When the lock is acquired before the action it should also be released after the action.

Reply:
No, we need to lock the action until the TTL expires, otherwise, the locking would be useless.

The whole point of using the lock is to make sure no other termination will happen until the TTL expires, its not about exclusive access

Reply:
Then I do not understand how the unlocking by expiration is synchronized with whatever action we want to have to happen while the lock is active. The point is that pure expiration is time based and time based systems are super hard to synchronize. This is why the mechanism should be event based so that the unlocking can be triggered somehow. So, what is the trigger?

Reply:
the unlock is triggered by time

 when a termination  should happen the  code will try to acquire lock  ->:
* if there is no lock or expired lock - > it will create a lock and proceed with termination
* if there is a lock that is not yet expired, it will skip the termination and it will try again in the next loop


The whole goal of this lock system is to ensure that we dont terminate many instances at once. Instead it will always have a 10 minute pause between killing nodes so there is enought time for the terminated node to properly boot and server pods.

Reply:
That there is "enough" time for anything at any given moment is not more than a wish in time based systems. Something will go wrong, the time will be up and the system will degrade eventually. Just saying, this design is wrong. It might be good enough for a year or two, until it will break and then everyone is wondering why and how this works. I will leave this now up to you because I think I made clear what is "right" or "wrong" IMO.

Reply:
I do not have a better ideas as having a way how to determine the terminated nodes are back in system and all customer pods that were on the terminated are somehow running and fill in sync. There is no such complex system that we could use or build in good timeframe. 

So we can drop this feature and focus on something else I guess.

Reply:
Go ahead when you think this is good enough. Just saying we should make conscious design decisions about time based and event based solutions. Let's at least document the workflow and the implication somehow. Because acquiring a lock and treat it as fire and forget raises more questions than it answers and will cause problems eventually. You may also want to consult @giantswarm/sig-operator for more opinions.

Reply:
ping @giantswarm/sig-infrastructure  for potential opinions as well

Reply:
The concept of lock is a bit misleading here, but I get the point and I guess it's one way to implement this. Perhaps a descriptive comment near the locking would be a good thing for maintenance purposes.

One alternative that comes to my mind is something similar to how cluster-autoscaler works:
- When faulty nodes are detected, tag the corresponding EC2 instances as such.
  - If there's need for detain period, one could have a monotonous counter as value, increased by each reconciliation loop that still finds the node unhealthy.
  - If node is not unhealthy, the tag would be removed.
- As a separate [later] step in reconciliation: When there are EC2 instances tagged as faulty (and potentially having counter value high enough), one of them is selected and terminated.
  - Again: If there's need for graceful period, one could reset other EC2 instances with faulty tag (i.e. remove it and start from scratch).

This approach shouldn't terminate nodes too quickly even if reconciliation loop is frequently run due to CR updates. The timing here is relative and could be adjusted by requiring e.g. 3 reconciliation runs w/ faulty node state before the node can be considered for termination.

WDYT?

Reply:
@tuommaki  This is exactly what the `badnodedection` package does, it tags unhealthy node with annotation, and each loop it increases the counter, at a certain point, it will mark this node for termination and will return it back to the operator, and then operator takes care of termination


But what I am talking about here is once the termination happens, it safer to a have graceful period before other termination happens. I thought that having a exact time for that is much easier readble than any other solutions. Resetting other counters to zero feels little bit hacky and harder to find IMHO.

If everyones finds this better than i will do it as it feels its not really any different than current solution, it jsut implements the `pause` on a little bit different counter. its still  'some time'.

Reply:
> it feels it's not really any different than the current solution, it just implements the pause on a little bit different counter. It's still 'some time'.

True. For overall picture it doesn't change much, but for certain corner cases (mainly clock drift) and potential first impression on out-of-context code block it might differ. Nevertheless, I'm fine either way.

Reply:
I have removed the lock and used ResetTickoOunters instead to clear all the counters on all nodes